From: Jan Blunck <jblunck@suse.de>
Subject: Reduce size of struct mem_cgroup_stat

Allocate struct mem_cgroup_stat->cpustat dynamically based on the
number of cpus configured at boot time rather than using the
compile-time maximum. (With 4096 cpus, this can save a lot of
memory.)

Signed-off-by: Andreas Gruenbacher <agruen@suse.de>

---
 mm/memcontrol.c |   25 ++++++++++++++++++++++---
 1 file changed, 22 insertions(+), 3 deletions(-)

Index: b/mm/memcontrol.c
===================================================================
--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -59,7 +59,7 @@ struct mem_cgroup_stat_cpu {
 } ____cacheline_aligned_in_smp;
 
 struct mem_cgroup_stat {
-	struct mem_cgroup_stat_cpu cpustat[NR_CPUS];
+	struct mem_cgroup_stat_cpu *cpustat;
 };
 
 /*
@@ -143,6 +143,7 @@ struct mem_cgroup {
 	struct mem_cgroup_stat stat;
 };
 static struct mem_cgroup init_mem_cgroup;
+static struct mem_cgroup_stat_cpu init_mem_cgroup_stat_cpu[NR_CPUS];
 
 /*
  * We use the lower bit of the page->page_cgroup pointer as a bit spin
@@ -1097,19 +1098,36 @@ static void free_mem_cgroup_per_zone_inf
 static struct mem_cgroup *mem_cgroup_alloc(void)
 {
 	struct mem_cgroup *mem;
+	struct mem_cgroup_stat_cpu *cpustat;
 
 	if (sizeof(*mem) < PAGE_SIZE)
 		mem = kmalloc(sizeof(*mem), GFP_KERNEL);
 	else
 		mem = vmalloc(sizeof(*mem));
 
-	if (mem)
-		memset(mem, 0, sizeof(*mem));
+	if (!mem)
+		goto out;
+	memset(mem, 0, sizeof(*mem));
+
+	cpustat = vmalloc(nr_cpu_ids * sizeof(*cpustat));
+	if (!cpustat)
+		goto out_mem;
+	memset(cpustat, 0, sizeof(*cpustat));
+	mem->stat.cpustat = cpustat;
 	return mem;
+out_mem:
+	if (sizeof(*mem) < PAGE_SIZE)
+		kfree(mem);
+	else
+		vfree(mem);
+out:
+	return NULL;
 }
 
 static void mem_cgroup_free(struct mem_cgroup *mem)
 {
+	vfree(mem->stat.cpustat);
+
 	if (sizeof(*mem) < PAGE_SIZE)
 		kfree(mem);
 	else
@@ -1125,6 +1143,7 @@ mem_cgroup_create(struct cgroup_subsys *
 
 	if (unlikely((cont->parent) == NULL)) {
 		mem = &init_mem_cgroup;
+		mem->stat.cpustat = &init_mem_cgroup_stat_cpu[0];
 		page_cgroup_cache = KMEM_CACHE(page_cgroup, SLAB_PANIC);
 	} else {
 		mem = mem_cgroup_alloc();
